src/Math/src/AlgebraicRiccatiEquation.cpp

```diff
@@ -0,0 +1,129 @@
+ * distributed under the terms of the GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#include <iostream>
+#include <sstream>
+
+#include <BipedalLocomotion/Math/AlgebraicRiccatiEquation.h>
+
+
+std::pair<bool, Eigen::MatrixXd>
+BipedalLocomotion::Math::ContinuousAlgebraicRiccatiEquation(Eigen::Ref<const Eigen::MatrixXd> A,
+                                                            Eigen::Ref<const Eigen::MatrixXd> B,
+                                                            Eigen::Ref<const Eigen::MatrixXd> Q,
+                                                            Eigen::Ref<const Eigen::MatrixXd> R,
+                                                            double tolerance,
+                                                            std::size_t maxIterations,
+                                                            bool verbose)
+{
+    constexpr std::string_view errorPrefix = "[BipedalLocomotion::Math::"
+                                             "ContinuousAlgebraicRiccatiEquation] ";
+
+    std::pair<bool, Eigen::MatrixXd> result;
+
+    const std::size_t n = B.rows();
+    const std::size_t m = B.cols();
+
+    // check the size of the matrices
+    if (A.rows() != n || A.cols() != n)
+    {
+        std::cerr << errorPrefix << "The matrix A must be a square matrix. Expected size: " << n
+                  << " x " << n << ". Passed size: " << A.rows() << " x " << A.cols() << "."
+                  << std::endl;
+        result.first = false;
+        return result;
+    }
+
+    if (Q.rows() != n || Q.cols() != n)
+    {
+        std::cerr << errorPrefix << "The matrix Q must be a square matrix. Expected size: " << n
+                  << " x " << n << ". Passed size: " << Q.rows() << " x " << Q.cols() << "."
+                  << std::endl;
+        result.first = false;
+        return result;
+    }
+
+    if (R.rows() != m || R.cols() != m)
+    {
+        std::cerr << errorPrefix << "The matrix Q must be a square matrix. Expected size: " << n
+                  << " x " << n << ". Passed size: " << Q.rows() << " x " << Q.cols() << "."
+                  << std::endl;
+        result.first = false;
+        return result;
+    }
+
+    Eigen::LLT<Eigen::MatrixXd> R_cholesky(R);
+    if (R_cholesky.info() != Eigen::Success)
+    {
+        std::cerr << errorPrefix << "The matrix R must be positive definite." << std::endl;
+        result.first = false;
+        return result;
+    }
+
+    constexpr double symmetricTolerance = 1e-8;
+    if (!Q.isApprox(Q.transpose(), symmetricTolerance))
+    {
+        std::cerr << errorPrefix << "The matrix Q must be symmetric." << std::endl;
+        result.first = false;
+        return result;
+    }
+
+    Eigen::MatrixXd H(2 * n, 2 * n);
+    H << A, B * R_cholesky.solve(B.transpose()),
+         Q, -A.transpose();
+
+    Eigen::MatrixXd Z = H;
+    Eigen::MatrixXd Z_old;
```

Comment:
I would prefer solutions that minimize the chance of memory allocation. Consider turning this as parameters of a class. In case you want to keep the function-like approach, you can still define a static function, but at the point is the user that decides to not care about memory allocation.